src/storage/buffer_manager/buffer_manager.cpp

```diff
@@ -33,61 +49,215 @@ void BufferManager::resize(uint64_t newSizeForDefaultPagePool, uint64_t newSizeF
+// or we can find no more pages to be evicted.
+// Lastly, we double check if the needed memory is available. If not, we free the memory we reserved
+// and return false, otherwise, we load the page to its corresponding frame and return true.
+bool BufferManager::claimAFrame(BufferManagedFileHandle& fileHandle, common::page_idx_t pageIdx,
+    PageReadPolicy pageReadPolicy) {
+    page_offset_t pageSizeToClaim = fileHandle.getPageSize();
+    // Reserve the memory for the page.
+    auto currentUsedMem = reserveUsedMemory(pageSizeToClaim);
+    EvictionCandidate evictionCandidate;
+    uint64_t claimedMemory = 0;
+    // Evict pages if necessary until we have enough memory.
+    while ((currentUsedMem + pageSizeToClaim - claimedMemory) > maxMemory.load()) {
+        if (!evictionQueue->dequeue(evictionCandidate)) {
+            // Cannot find more pages to be evicted. Free the memory we reserved and return false.
+            freeUsedMemory(pageSizeToClaim);
+            return false;
+        }
+        if (evictionCandidate.evictionTimestamp !=
+                evictionCandidate.pageState->getEvictionTimestamp() ||
+            evictionCandidate.pageState->getPinCount() > 0) {
+            continue;
+        }
+        // Now we find a block whose pin count can be 0, and potentially haven't been accessed since
+        // enqueued. Try to lock the page and double check if the page's pin count is actually 0 and
+        // the page has not been accessed recently, if so, evict the page from its frame.
+        claimedMemory += tryUnloadPageFromFrame(evictionCandidate);
+        currentUsedMem = usedMemory.load();
+    }
+    if ((currentUsedMem + pageSizeToClaim - claimedMemory) > maxMemory.load()) {
+        // Cannot claim the memory needed. Free the memory we reserved and return false.
+        freeUsedMemory(pageSizeToClaim);
+        return false;
+    }
+    // Have enough memory available now, load the page into its corresponding frame.
+    loadPageIntoFrame(fileHandle, pageIdx, pageReadPolicy);
+    freeUsedMemory(claimedMemory);
+    return true;
+}
+
+void BufferManager::addToEvictionQueue(PageState* pageState) {
+    auto timestampBeforeEvict = pageState->incrementEvictionTimestamp();
+    if (++numEvictionQueueInsertions == BufferPoolConstants::PURGE_EVICTION_QUEUE_INTERVAL) {
+        evictionQueue->purge();
+        numEvictionQueueInsertions = 0;
+    }
+    evictionQueue->enqueue(pageState, timestampBeforeEvict + 1);
+}
+
+uint64_t BufferManager::tryUnloadPageFromFrame(EvictionCandidate& candidate) {
+    auto& pageState = *candidate.pageState;
+    if (!pageState.acquireLock(LockMode::NON_BLOCKING)) {
+        return 0;
+    }
+    // We check pinCount and evictionTimestamp again after acquiring the lock on page currently
+    // residing in the frame. At this point in time, no other thread can change the pinCount and the
+    // evictionTimestamp.
+    if (0u != pageState.getPinCount() ||
```

Comment:
Ditto about replacing this check with a utility function. This is the 3rd place the condition appears.